# Анализ продаж автомобилей в Казахстане в 2019 году

## Загрузка данных и просмотр общей информации

In [164]:
import pandas as pd
import re

In [165]:
df = pd.read_csv('autokz2019.csv',decimal =',', sep=';', skipinitialspace=True)

In [166]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39966 entries, 0 to 39965
Data columns (total 25 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Год                             39966 non-null  int64  
 1   Месяц                           39966 non-null  object 
 2   Компания                        39966 non-null  object 
 3   Бренд                           39966 non-null  object 
 4   Модель                          39966 non-null  object 
 5   Модификация                     36375 non-null  object 
 6   Год выпуска                     39465 non-null  object 
 7   Страна-производитель            39966 non-null  object 
 8   Вид топлива                     36826 non-null  object 
 9   Объём двиг, л,                  35708 non-null  object 
 10  Коробка передач                 36711 non-null  object 
 11  Тип привода                     35677 non-null  object 
 12  Сегмент                         

удаляем столбцы: наименование дилерского центра, форма расчета, сегмент, сегментация eng, локализация производства, тип клиента и модификация (они либо бесполезны, либо много пропусков)

In [167]:
df = df.drop(df.columns[[5, 12, 14, 15, 16, 23, 24]], axis=1)

In [168]:
df.head()

Год   Месяц     Компания Бренд Модель Год выпуска Страна-производитель  \
0  2019     Май  Mercur Auto  Audi     A3        2018             Германия   
1  2019  Август  Mercur Auto  Audi     A3        2018             Германия   
2  2019  Апрель  Mercur Auto  Audi     A4        2018             Германия   
3  2019    Июль  Mercur Auto  Audi     A4        2018             Германия   
4  2019    Июль  Mercur Auto  Audi     A4        2018             Германия   

  Вид топлива Объём двиг, л, Коробка передач Тип привода  Регион  Количество  \
0      Бензин            1,4        S-tronic    передний  Алматы         1.0   
1      Бензин            1,4        S-tronic    передний  Алматы         1.0   
2      Бензин            1,4        S-Tronic         FWD  Алматы         1.0   
3      Бензин            1,4        S-tronic    передний  Алматы         1.0   
4      Бензин            1,4        S-tronic    передний  Алматы         1.0   

   Цена, USD  Продажа, USD   Область     Сегментация 2013 Класс 2013  
0   28115.00      28115.00  г.Алматы  Легковые автомобили    C класс  
1   32246.99      32246.99  г.Алматы  Легковые автомобили    C класс  
2   32000.00      32000.00  г.Алматы  Легковые автомобили    D класс  
3   31929.00      31929.00  г.Алматы  Легковые автомобили    D класс  
4   31929.00      31929.00  г.Алматы  Легковые автомобили    D класс

Изменим наименования столбцов

In [169]:
df=df.rename(columns={
    'Год': 'year', 'Месяц': 'month', 'Компания': 'company', 'Бренд': 'brand', 'Модель': 'model', 'Год выпуска': 'year_of_release', \
    'Страна-производитель': 'country_of_manufacture', 'Вид топлива': 'fuel', 'Объём двиг, л,': 'engine_volume', 'Коробка передач': 'transmission_box', \
    'Тип привода': 'type_of_drive', 'Регион': 'region', 'Количество': 'quantity', 'Цена, USD': 'price', 'Продажа, USD': 'sale', \
    'Область': 'oblact', 'Сегментация 2013': 'segment', 'Класс 2013': 'class'})

In [170]:
df.head()

year   month      company brand model year_of_release  \
0  2019     Май  Mercur Auto  Audi    A3            2018   
1  2019  Август  Mercur Auto  Audi    A3            2018   
2  2019  Апрель  Mercur Auto  Audi    A4            2018   
3  2019    Июль  Mercur Auto  Audi    A4            2018   
4  2019    Июль  Mercur Auto  Audi    A4            2018   

  country_of_manufacture    fuel engine_volume transmission_box type_of_drive  \
0               Германия  Бензин           1,4         S-tronic      передний   
1               Германия  Бензин           1,4         S-tronic      передний   
2               Германия  Бензин           1,4         S-Tronic           FWD   
3               Германия  Бензин           1,4         S-tronic      передний   
4               Германия  Бензин           1,4         S-tronic      передний   

   region  quantity     price      sale    oblact              segment  \
0  Алматы       1.0  28115.00  28115.00  г.Алматы  Легковые автомобили   
1  Алматы       1.0  32246.99  32246.99  г.Алматы  Легковые автомобили   
2  Алматы       1.0  32000.00  32000.00  г.Алматы  Легковые автомобили   
3  Алматы       1.0  31929.00  31929.00  г.Алматы  Легковые автомобили   
4  Алматы       1.0  31929.00  31929.00  г.Алматы  Легковые автомобили   

     class  
0  C класс  
1  C класс  
2  D класс  
3  D класс  
4  D класс

## Предобработка данных

### Проверка на дубликаты

In [171]:
# проверим датафрейм на наличие явных дубликатов
df.duplicated().sum()

23471

Наличие большого количества дубликатов обусловлено тем, что данные представлены по месяцам. Скорее всего, продажи были проведены в разные дни. Дубли оставляем 

### Проверка на пропуски 

In [172]:
report = df.isna().sum().to_frame()
report = report.rename(columns = {0: 'missing_values'})
report['% of total'] = ((report['missing_values'] / df.shape[0])*100).round(2)
report.sort_values(by = 'missing_values', ascending = False)

missing_values  % of total
type_of_drive                     4289       10.73
engine_volume                     4258       10.65
transmission_box                  3255        8.14
fuel                              3140        7.86
year_of_release                    501        1.25
quantity                             6        0.02
region                               0        0.00
segment                              0        0.00
oblact                               0        0.00
sale                                 0        0.00
price                                0        0.00
year                                 0        0.00
month                                0        0.00
country_of_manufacture               0        0.00
model                                0        0.00
brand                                0        0.00
company                              0        0.00
class                                0        0.00

In [173]:
df[df['year_of_release'].isna()]

year     month                        company                brand  \
65     2019      Июнь              Автоцентр-Бавария                  BMW   
9599   2019       Май  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
9600   2019       Май  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
9601   2019       Май  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
9602   2019       Май  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
...     ...       ...                            ...                  ...   
39542  2019  Сентябрь           Volkswagen Group Rus           Volkswagen   
39554  2019      Июнь  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
39555  2019      Июнь  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
39556  2019      Июнь  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   
39557  2019      Июль  Hyundai Com Trans Kazakhstan   Hyundai Truck & Bus   

          model year_of_release country_of_manufacture    fuel engine_volume  \
65      7 серия             NaN               Германия  Бензин             3   
9599       H350             NaN   Республика Казахстан  Дизель           NaN   
9600       H350             NaN   Республика Казахстан  Дизель           NaN   
9601       H350             NaN   Республика Казахстан  Дизель           NaN   
9602       H350             NaN   Республика Казахстан  Дизель           NaN   
...         ...             ...                    ...     ...           ...   
39542  Teramont             NaN   Российская Федерация     NaN           NaN   
39554       EX8             NaN   Республика Казахстан     NaN           NaN   
39555       EX8             NaN   Республика Казахстан     NaN           NaN   
39556       EX8             NaN   Республика Казахстан     NaN           NaN   
39557       EX8             NaN   Республика Казахстан     NaN           NaN   

      transmission_box type_of_drive            region  quantity        price  \
65                АКПП        Полный  Усть-Каменогорск       1.0  93520.28222   
9599               NaN           2WD            Алматы       1.0  33833.35184   
9600               NaN           2WD            Алматы       1.0  33833.35184   
9601               NaN           2WD            Алматы       1.0  33833.35184   
9602               NaN           2WD            Алматы       1.0  33833.35184   
...                ...           ...               ...       ...          ...   
39542              NaN           NaN          Костанай       1.0  49655.99915   
39554              NaN           NaN            Алматы       1.0  34590.76936   
39555              NaN           NaN            Алматы       1.0  34590.76936   
39556              NaN           NaN            Алматы       1.0  34590.76936   
39557              NaN           NaN            Алматы       1.0  34431.63774   

              sale                          oblact                  segment  \
65     93520.28222  Восточно-Казахстанская область      Легковые автомобили   
9599   33833.35184                        г.Алматы  Коммерческие автомобили   
9600   33833.35184                        г.Алматы  Коммерческие автомобили   
9601   33833.35184                        г.Алматы  Коммерческие автомобили   
9602   33833.35184                        г.Алматы  Коммерческие автомобили   
...            ...                             ...                      ...   
39542  49655.99915            Костанайская область             Внедорожники   
39554  34590.76936                        г.Алматы  Коммерческие автомобили   
39555  34590.76936                        г.Алматы  Коммерческие автомобили   
39556  34590.76936                        г.Алматы  Коммерческие автомобили   
39557  34431.63774                        г.Алматы  Коммерческие автомобили   

                           class  
65                       F класс  
9599   Крупнотоннажные грузовики  
9600   Крупнотоннажные грузовики  
9601   Крупнотоннажные грузовики  
9602   Крупнотоннажные г

Так как количество пропусков в столбцах 'quantity' и'year_of_release' невелико, удалим их из датасета

In [174]:
df=df.dropna(axis=0, subset=['quantity','year_of_release'])

## Обработка данных в столбцах

### Год выпуска автомобиля

In [175]:
# посмотрим уникальные значения
df['year_of_release'].unique()

array(['2018', '2017', '2019', '2016', '2014', '2013', '2011', '2\xa0018',
       '2\xa0019'], dtype=object)

In [176]:
# скомпилируем регулярные выражения, чтобы ускорить их применение. Здесь '\D' удаляет все символы, не являющиеся цифрами, в столбце 'year_of_release'.

magic_wand = re.compile(r'\D') 
df['year_of_release'] = df['year_of_release'].map(lambda x: magic_wand.sub('', x))

In [177]:
df_release=df['year_of_release'].value_counts(dropna=False).to_frame(name='Количество')
df_release['Процент (%)'] = (df_release['Количество'] / len(df) * 100).round(2)

In [178]:
print("\nСтатистика по 'year_of_release':")
print(df_release.head(10))


Статистика по 'year_of_release':
                 Количество  Процент (%)
year_of_release                         
2019                  30023        76.09
2018                   9078        23.01
2017                    324         0.82
2016                     30         0.08
2014                      2         0.01
2013                      1         0.00
2011                      1         0.00


На авторынке Казахстана в 2019 году 76% проданных авто - 2019 года выпуска, 23% - 2018 года выпуска. 

### Наименования компаний-дилеров

In [179]:
# посмотрим уникальные значения
df['company'].unique()

array(['Mercur Auto', 'Автоцентр-Бавария', 'БИПЭК АВТО', 'Вираж',
       'Eurasia Motor Premium', 'Daewoo Bus Kazakhstan', 'Caspian Motors',
       'Каспиан Моторс', 'СемАЗ', 'Astana Motors', 'Astana Motors ',
       'Hyundai Com Trans Kazakhstan ', 'Nissan Manufacturing RUS',
       'СВС-ТРАНС', 'Allur Auto', 'Almaty Motors Premium', 'TERRA MOTORS',
       'ТК КАМАЗ', 'Toyota Motor Kazakhstan', 'Лифан Моторс Рус',
       'MAN Truck & Bus Kazakhstan', 'Автомир ГК', 'Autokapital',
       'Автокапитал', 'ММС Рус', 'MMC RUS', 'Mercur Autos',
       'УзАвто-Казахстан', 'Равон Моторс Казахстан',
       'Ravon Motors Kazakstan', 'Renault Россия', 'Scania Central Asia',
       'Subaru Kazakhstan', 'Volkswagen Group Rus', 'Автодом Motors KST',
       'Автомир-Центр', 'ТОО "Eurasia Motor Zhaik"',
       'Scandinavian Motors', 'Хино Моторс Казахстан', 'Hino Motors '],
      dtype=object)

In [180]:
# приводим всё к строке и удаляем лишние пробелы
df['company'] = df['company'].astype(str).str.strip()

По данным сайта 2gis.kz, наименования 'Mercur Autos' и 'Mercur Auto', 'Автомир ГК' и 'Автомир-Центр' относятся к одной и той же компании. Также дублируются наименования на русском и английском языках: 'Хино Моторс Казахстан' и 'Hino Motors', 'Каспиан Моторс' и 'Caspian Motors', 'Автокапитал' и 'Autokapital'.
Изменим значения столбца 'company'

In [181]:
df['company']= df['company'].replace({'Хино Моторс Казахстан':'Hino Motors', 'Каспиан Моторс':'Caspian Motors', 'Автокапитал':'Autokapital', 'Mercur Autos':'Mercur Auto', 'Автомир ГК':'Автомир-Центр'})

In [182]:
df_company=df['company'].value_counts(dropna=False).to_frame(name='Количество')
df_company['Процент (%)'] = (df_company['Количество'] / len(df) * 100).round(2)

In [183]:
print("\nСтатистика по 'company':")
print(df_company)


Статистика по 'company':
                              Количество  Процент (%)
company                                              
Toyota Motor Kazakhstan            11951        30.29
Astana Motors                      10169        25.77
БИПЭК АВТО                          3618         9.17
Вираж                               2909         7.37
Allur Auto                          2305         5.84
Renault Россия                      1796         4.55
Nissan Manufacturing RUS            1149         2.91
Равон Моторс Казахстан               947         2.40
ТК КАМАЗ                             893         2.26
Mercur Auto                          643         1.63
Ravon Motors Kazakstan               559         1.42
Subaru Kazakhstan                    503         1.27
ММС Рус                              362         0.92
Лифан Моторс Рус                     233         0.59
Автоцентр-Бавария                    223         0.57
СВС-ТРАНС                            202         0.51
Уз

In [184]:
# посмотрим уникальные значения
df['brand'].unique()

array(['Audi', 'BMW', 'Chevrolet', 'Daewoo', 'Ford', 'Foton', 'GAZ',
       'Hyundai', 'Hyundai Truck & Bus', 'Infiniti', 'Isuzu', 'Iveco',
       'Jac', 'Jaguar', 'KAMAZ', 'Kia', 'Lada', 'Land Rover', 'Lexus',
       'Lifan', 'MAN', 'Mazda', 'Mercedes-Benz', 'Mercedes-Benz Trucks',
       'Mini', 'Mitsubishi', 'Nissan', 'Peugeot', 'Porsche', 'Ravon',
       'Renault', 'Scania', 'Skoda', 'Subaru', 'Toyota', 'UAZ',
       'Volkswagen', 'Volvo', 'Урал', 'Cadillac', 'Shacman', 'Hino',
       'Dong Feng', 'ANKAI', 'Nefaz', 'Yutong'], dtype=object)

In [185]:
df_brand=df['brand'].value_counts(dropna=False).to_frame(name='Количество')
df_brand['Процент (%)'] = (df_brand['Количество'] / len(df) * 100).round(2)

In [186]:
print("\nСтатистика по 'brand':")
print(df_brand)


Статистика по 'brand':
                      Количество  Процент (%)
brand                                        
Toyota                     10745        27.23
Hyundai                    10168        25.77
Lada                        1967         4.98
GAZ                         1821         4.61
Renault                     1796         4.55
Ravon                       1784         4.52
Jac                         1770         4.49
Lexus                       1206         3.06
Nissan                      1053         2.67
UAZ                          925         2.34
KAMAZ                        892         2.26
Kia                          831         2.11
Volkswagen                   682         1.73
Chevrolet                    635         1.61
Subaru                       504         1.28
Skoda                        378         0.96
Mitsubishi                   364         0.92
ANKAI                        300         0.76
Lifan                        233         0.59
BMW       

In [187]:
df['model'].unique()

array(['A3', 'A4', 'A6', 'A7', 'A8', 'Q5', 'Q7', '3 серия', '5 серия',
       '6 серия', '7 серия', 'X1', 'X3', 'X4', 'X5', 'X6', 'X7', 'Z4',
       'Aveo', 'Captiva', 'Niva', 'Tahoe', 'Tracker', 'BS 106 A',
       'Fiesta', 'Mondeo', 'BJ3253DMPKB-AD', '2217', '2310', '2705',
       '2752', '3221', '3302', '3308', '3309', 'Next', 'Accent', 'Creta',
       'Elantra', 'Santa Fe', 'Sonata', 'TQ-1/H-1', 'County', 'H350',
       'HD 170', 'HD 65', 'HD 78', 'Universe', 'Tucson', 'Q50', 'QX50',
       'QX60', 'QX80', 'D-MAX', 'NMR', 'NPR', 'Daily', 'Daily A',
       'Daily V', 'Stralis', 'Tipper', 'S3', 'S5', 'F-Pace', 'F-Type',
       'XE', 'XF', 'XJ', '4308', '4311', '43118', '43253', '43502',
       '44108', '45142', '45143', '53215', '53504', '53605', '54115',
       '5490', '65111', '65115', '65116', '65117', '6520', '65206',
       '65225', '66052', '6606', 'Cerato', 'Optima', 'Picanto', 'Quoris',
       'Rio', 'Sorento', 'Soul', 'Sportage', '4x4', 'Granta', 'Kalina',
       'Largus', '

In [188]:
df['country_of_manufacture'].unique()

array(['Германия', 'США', 'Австрия', 'Республика Казахстан',
       'Российская Федерация', 'Корея', 'Япония', 'Таиланд', 'Китай',
       'UK', 'Узбекистан', 'Венгрия', 'Турция', 'Испания', 'Нидерланды',
       'Польша', 'Швеция', 'Белоруссия', 'Бельгия'], dtype=object)

кодируем страны в столбце 'страна производства' в alpha3 (RUS, KAZ, USA и т.д.)


In [189]:
df['country_of_manufacture'] = df['country_of_manufacture'].replace({
    'Германия': 'DEU',
    'США': 'USA',
    'Австрия': 'AUT', 
    'Республика Казахстан': 'KAZ',
    'Российская Федерация': 'RUS',
    'Корея': 'KOR',
    'Япония': 'JPN',
    'Таиланд': 'THA',
    'Китай': 'CHN',
    'UK': 'GBR',
    'Узбекистан': 'UZB',
    'Венгрия': 'HUN',
    'Турция': 'TUR',
    'Испания': 'ESP',
    'Нидерланды': 'NLD',
    'Польша': 'POL',
    'Швеция': 'SWE', 
    'Белоруссия': 'BLR', 
    'Бельгия': 'BEL'
})    

In [190]:
df_country=df['country_of_manufacture'].value_counts(dropna=False).to_frame(name='Количество')
df_country['Процент (%)'] = (df_country['Количество'] / len(df) * 100).round(2)

In [191]:
print("\nСтатистика по 'country_of_manufacture':")
print(df_country)


Статистика по 'country_of_manufacture':
                        Количество  Процент (%)
country_of_manufacture                         
KAZ                          19038        48.25
RUS                          12702        32.19
JPN                           4591        11.63
TUR                           1167         2.96
THA                            959         2.43
USA                            289         0.73
DEU                            257         0.65
GBR                            190         0.48
CHN                             94         0.24
UZB                             70         0.18
AUT                             33         0.08
KOR                             17         0.04
SWE                             16         0.04
NLD                             12         0.03
ESP                              9         0.02
POL                              6         0.02
BEL                              6         0.02
HUN                              2         0.01

### Обработка значений столбца 'тип топлива'

In [192]:
df['fuel'].unique()

array(['Бензин', 'Дизель', nan, 'дизель ', 'бензин ', 'гибрид', '2',
       '1,6', '0', 'Электро', 'Электричество'], dtype=object)

In [193]:
df[df['fuel']=='2']

year   month         company    brand   model year_of_release  \
23723  2019  Январь  Renault Россия  Renault  Duster            2019   
23724  2019  Январь  Renault Россия  Renault  Duster            2018   
23725  2019  Январь  Renault Россия  Renault  Duster            2018   
23726  2019  Январь  Renault Россия  Renault  Duster            2018   
23727  2019  Январь  Renault Россия  Renault  Duster            2018   
23729  2019  Январь  Renault Россия  Renault  Duster            2018   
23730  2019  Январь  Renault Россия  Renault  Duster            2018   
23731  2019  Январь  Renault Россия  Renault  Duster            2018   
23732  2019  Январь  Renault Россия  Renault  Duster            2018   
23746  2019  Январь  Renault Россия  Renault  Duster            2018   
23747  2019  Январь  Renault Россия  Renault  Duster            2018   
23748  2019  Январь  Renault Россия  Renault  Duster            2018   
23749  2019  Январь  Renault Россия  Renault  Duster            2018   
23750  2019  Январь  Renault Россия  Renault  Duster            2018   
23752  2019  Январь  Renault Россия  Renault  Duster            2018   
23753  2019  Январь  Renault Россия  Renault  Duster            2018   
23755  2019  Январь  Renault Россия  Renault  Duster            2018   
23758  2019  Январь  Renault Россия  Renault  Duster            2018   
23759  2019  Январь  Renault Россия  Renault  Duster            2018   
23760  2019  Январь  Renault Россия  Renault  Duster            2018   
23761  2019  Январь  Renault Россия  Renault  Duster            2018   
23762  2019  Январь  Renault Россия  Renault  Duster            2018   
23765  2019  Январь  Renault Россия  Renault  Duster            2018   

      country_of_manufacture fuel engine_volume transmission_box  \
23723                    RUS    2            MT              4WD   
23724                    RUS    2            AT              4WD   
23725                    RUS    2            AT              4WD   
23726                    RUS    2            AT              4WD   
23727                    RUS    2            MT              4WD   
23729                    RUS    2            AT              4WD   
23730                    RUS    2            AT              4WD   
23731                    RUS    2            MT              4WD   
23732                    RUS    2            MT              4WD   
23746                    RUS    2            AT              4WD   
23747                    RUS    2            MT              4WD   
23748                    RUS    2            AT              4WD   
23749                    RUS    2            AT              4WD   
23750                    RUS    2            AT              4WD   
23752                    RUS    2            AT              4WD   
23753                    RUS    2            AT              4WD   
23755                    RUS    2            MT              4WD   
23758                    RUS    2            AT              4WD   
23759                    RUS    2            AT              4WD   
23760                    RUS    2            AT              4WD   
23761                    RUS    2            AT              4WD   
23762                    RUS    2            AT              4WD   
23765                    RUS    2            MT              4WD   

      type_of_drive            region  quantity        price         sale  \
23723           4WD           Уральск       1.0  12249.61125  12249.61125   
23724           4WD        Нур-Султан       1.0  12249.61125  12249.61125   
23725           4WD        Нур-Султан       1.0  12249.61125  12249.61125   
23726           4WD         Кызылорда       1.0  12249.61125  12249.61125   
23727           4WD         Караганда       1.0  12249.61125  12249.61125   
23729           4WD         Караганда       1.0  12249.61125  12249.61125   
23730           4WD         Караганда       1.0  12249.61125  12249.61125   
23731           4WD         К

In [194]:
df[df['fuel']=='2']['fuel'] = 'F'

C:\Users\111\AppData\Local\Temp\ipykernel_15156\1757200051.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[df['fuel']=='2']['fuel'] = 'F'


In [195]:
df[df['fuel']=='1,6']

year   month         company    brand    model year_of_release  \
23728  2019  Январь  Renault Россия  Renault   Duster            2018   
23733  2019  Январь  Renault Россия  Renault   Duster            2018   
23734  2019  Январь  Renault Россия  Renault   Duster            2018   
23735  2019  Январь  Renault Россия  Renault   Duster            2018   
23736  2019  Январь  Renault Россия  Renault   Duster            2018   
23751  2019  Январь  Renault Россия  Renault   Duster            2018   
23754  2019  Январь  Renault Россия  Renault   Duster            2018   
23763  2019  Январь  Renault Россия  Renault   Duster            2018   
23764  2019  Январь  Renault Россия  Renault   Duster            2018   
24997  2019  Январь  Renault Россия  Renault  Sandero            2018   
24998  2019  Январь  Renault Россия  Renault  Sandero            2018   
24999  2019  Январь  Renault Россия  Renault  Sandero            2018   
25002  2019  Январь  Renault Россия  Renault  Sandero            2018   
25023  2019  Январь  Renault Россия  Renault  Sandero            2018   

      country_of_manufacture fuel engine_volume transmission_box  \
23728                    RUS  1,6            MT              4WD   
23733                    RUS  1,6            MT              4WD   
23734                    RUS  1,6            AT         Передний   
23735                    RUS  1,6            AT         Передний   
23736                    RUS  1,6            AT         Передний   
23751                    RUS  1,6            AT         Передний   
23754                    RUS  1,6            MT              4WD   
23763                    RUS  1,6            MT              4WD   
23764                    RUS  1,6            MT              4WD   
24997                    RUS  1,6             0                0   
24998                    RUS  1,6            MT         Передний   
24999                    RUS  1,6            MT         Передний   
25002                    RUS  1,6             0                0   
25023                    RUS  1,6             0                0   

      type_of_drive            region  quantity        price         sale  \
23728           4WD         Караганда       1.0  12249.61125  12249.61125   
23733           4WD         Караганда       1.0  12249.61125  12249.61125   
23734      Передний         Караганда       1.0  12249.61125  12249.61125   
23735      Передний         Караганда       1.0  12249.61125  12249.61125   
23736      Передний         Караганда       1.0  12249.61125  12249.61125   
23751      Передний  Усть-Каменогорск       1.0  12249.61125  12249.61125   
23754           4WD            Алматы       1.0  12249.61125  12249.61125   
23763           4WD        Нур-Султан       1.0  12249.61125  12249.61125   
23764           4WD     Петропавловск       1.0  12249.61125  12249.61125   
24997             0            Атырау       1.0  12476.89637  12476.89637   
24998      Передний           Уральск       1.0  12476.89637  12476.89637   
24999      Передний           Уральск       1.0  12476.89637  12476.89637   
25002             0            Атырау       1.0  12476.89637  12476.89637   
25023             0        Нур-Султан       1.0  12476.89637  12476.89637   

                               oblact              segment              class  
23728          Карагандинская область         Внедорожники  Субкомпактные SUV  
23733          Карагандинская область         Внедорожники  Субкомпактные SUV  
23734          Карагандинская область         Внедорожники  Субкомпактные SUV  
23735          Карагандинская область         Внедорожники  Субкомпактные SUV  
23736          Карагандинская область         Внедорожники  Субкомпактные SUV  
23751  Восточно-Казахстанская область         Внедорожники  Субкомпактные SUV  
23754                        г.Алматы         Внедорожники  Субкомпактные SUV  
23763                    г.Нур-Султан         Внедорожники  Субкомпактные SUV  
23764    Север

In [196]:
df[df['fuel']=='1,6']['fuel'] = 'F'

C:\Users\111\AppData\Local\Temp\ipykernel_15156\1090302158.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[df['fuel']=='1,6']['fuel'] = 'F'


In [197]:
df[df['fuel']=='0']

year   month     company brand model year_of_release  \
37121  2019  Январь  БИПЭК АВТО   UAZ  3741            2018   
37122  2019  Январь  БИПЭК АВТО   UAZ  3741            2018   
37123  2019  Январь  БИПЭК АВТО   UAZ  3741            2018   

      country_of_manufacture fuel engine_volume transmission_box  \
37121                    KAZ    0         2,693             5 МТ   
37122                    KAZ    0         2,693             5 МТ   
37123                    KAZ    0         2,693             5 МТ   

      type_of_drive       region  quantity    price     sale  \
37121        Полный  Талдыкорган       1.0  14200.0  14200.0   
37122        Полный       Актобе       3.0  14200.0  42600.0   
37123        Полный        Актау       2.0  14200.0  28400.0   

                      oblact                  segment                    class  
37121    Алматинская область  Коммерческие автомобили  Малотоннажные грузовики  
37122    Актюбинская область  Коммерческие автомобили  Малотоннажные грузовики  
37123  Мангистауская область  Коммерческие автомобили  Малотоннажные грузовики

In [198]:
df[df['fuel']=='0']['fuel'] = 'F'

C:\Users\111\AppData\Local\Temp\ipykernel_15156\3070535474.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[df['fuel']=='0']['fuel'] = 'F'


Renault Duster 1.6 и 2.0 2018 и 2019 года выпуска - с бензиновым двигателем, UAZ 3741 - также работает на бензине

In [199]:
df['fuel']=df['fuel'].fillna('unknown')

Кодируем тип топлива в (F - fuel, D - diesel, E - electro, HYB - hybrid) и меняем значения в столбце. Пропуски заменим на "неизвестно"


In [200]:
df['fuel'] = df['fuel'].astype(str).str.strip()

In [201]:
df['fuel'] = df['fuel'].str.lower()

In [202]:
# 
df['fuel'] = df['fuel'].replace({'бензин': 'F', '0':'F','1,6':'F','2':'F','дизель': 'D', 'электро': 'E', 'электричество':'E', 'гибрид': 'HYB'})

In [203]:
df['fuel'].unique()

array(['F', 'D', 'unknown', 'HYB', 'E'], dtype=object)

### Тип привода

В автомобилях бывают три основных типа приводов:

1. Передний привод (FWD, Front wheel drive). 
2. Задний привод (RWD, Rear Wheel Drive). 
3. Полный привод (4WD, Four wheel drive). Имеет разновидности:
   - Подключаемый полный привод (Part Time).
   - Постоянный полный привод (Full Time).
   - Автоматически подключаемый полный привод (AWD, All wheel drive).

In [204]:
df['type_of_drive'].unique()

array(['передний', 'FWD', 'quattro', 'Задний', 'Полный', 'AWD',
       'Передний', '4WD', '4х2.2', nan, 'RWD', '2 WD', '2WD', '4 WD',
       '4X4', '4х4', '4x4', '4x2', '4х2', '4X2', 'Астана',
       'Передний (FF)', '0', 'FF', '#Н/Д', 'полный', '4Motion'],
      dtype=object)

В автомобилях бывают 4 основных типа приводов:
    Передний привод (FWD, Front wheel drive). 
    Задний привод (RWD, Rear Wheel Drive).
    Полный привод (4WD, Four wheel drive).
    Автоматически подключаемый полный привод (AWD, All wheel drive).
    Типы привода 2WD:
Передний привод (FWD). В таких автомобилях крутящий момент от двигателя передаётся на передние колёса, в то время как задние вращаются свободно. 
dubizzle.com
avtobeginner.ru
drive2.ru
Задний привод (RWD). 

    quattro — название, используемое компанией AUDI AG для обозначения технологий, либо систем постоянного полного привода. 4Motion (4MOTION) — фирменная система полного привода автомобилей Volkswagen.
    #Н/Д - нет данных
    4x2 или 2WD имеют привод на два колеса (2WD) с четырьмя колесами (
    Автомобиль КамАЗ 65116 имеет задний (FR) привод.
    КамАЗ 45141 имеет полный привод (4WD)
    Автомобиль Renault Duster может иметь следующие типы привода: передний (FF) и полный (4WD). 
    Все версии автомобиля Renault Sandero — с передним приводом


In [205]:
df[df['type_of_drive']=='Астана']

year month   company  brand  model year_of_release  \
15771  2019   Май  ТК КАМАЗ  KAMAZ  65116            2019   
39549  2019   Май  ТК КАМАЗ  KAMAZ  45141            2019   

      country_of_manufacture fuel engine_volume transmission_box  \
15771                    KAZ    D           NaN               MT   
39549                    KAZ    D           NaN               MT   

      type_of_drive      region  quantity        price         sale  \
15771        Астана  Нур-Султан       1.0  55655.86316  55655.86316   
39549        Астана  Нур-Султан       1.0  53963.81966  53963.81966   

             oblact                  segment                      class  
15771  г.Нур-Султан  Коммерческие автомобили  Крупнотоннажные грузовики  
39549  г.Нур-Султан  Коммерческие автомобили  Крупнотоннажные грузовики

In [206]:
df.loc[15771,'type_of_drive'] = 'RWD'

In [207]:
df.loc[39549,'type_of_drive'] = '4WD'

In [208]:
df['type_of_drive']=df['type_of_drive'].astype(str)

In [209]:
df['type_of_drive']=df['type_of_drive'].str.upper()

In [213]:
# 
def drive_type(value):
    if value in ['ПЕРЕДНИЙ','ПЕРЕДНИЙ (FF)','FF', 'FWD']:
        return 'FWD'
    elif value in ['ЗАДНИЙ','RWD']:
        return 'RWD'
    elif '2' in value:
        return '2WD'
    elif 'AWD' in value:
        return value    
    elif value in ['QUATTRO', '4MOTION', '4WD', '4 WD', '4X4', '4Х4', 'ПОЛНЫЙ']: 
        return '4WD'
    return 'unknown'

In [214]:
df['type_of_drive'] = df['type_of_drive'].apply(drive_type)

In [217]:
df['type_of_drive'].value_counts(normalize=True)

type_of_drive
4WD        0.355255
2WD        0.248131
FWD        0.228921
unknown    0.100864
RWD        0.042449
AWD        0.024380
Name: proportion, dtype: float64

In [216]:
df[df['type_of_drive']=='unknown']

year     month   company  brand           model year_of_release  \
861    2019      Март     СемАЗ  Foton  BJ3253DMPKB-AD            2019   
862    2019      Июнь     СемАЗ  Foton  BJ3253DMPKB-AD            2018   
863    2019      Июнь     СемАЗ  Foton  BJ3253DMPKB-AD            2018   
864    2019      Июль     СемАЗ  Foton  BJ3253DMPKB-AD            2018   
865    2019    Август     СемАЗ  Foton  BJ3253DMPKB-AD            2018   
...     ...       ...       ...    ...             ...             ...   
39961  2019  Сентябрь  ТК КАМАЗ  KAMAZ          58815Z            2019   
39962  2019  Сентябрь  ТК КАМАЗ  KAMAZ          58815Z            2019   
39963  2019  Сентябрь  ТК КАМАЗ  KAMAZ          58815Z            2019   
39964  2019  Сентябрь  ТК КАМАЗ  KAMAZ          58815Z            2019   
39965  2019  Сентябрь  ТК КАМАЗ  KAMAZ          58815Z            2019   

      country_of_manufacture     fuel engine_volume transmission_box  \
861                      KAZ  unknown           NaN              NaN   
862                      KAZ  unknown           NaN              NaN   
863                      KAZ  unknown           NaN              NaN   
864                      KAZ  unknown           NaN              NaN   
865                      KAZ  unknown           NaN              NaN   
...                      ...      ...           ...              ...   
39961                    KAZ        D           NaN               MT   
39962                    KAZ        D           NaN               MT   
39963                    KAZ        D           NaN               MT   
39964                    KAZ        D           NaN               MT   
39965                    KAZ        D           NaN               MT   

      type_of_drive            region  quantity        price          sale  \
861         unknown            Алматы       1.0  48275.86207   48275.86207   
862         unknown  Усть-Каменогорск       1.0  49735.91972   49735.91972   
863         unknown          Кокшетау       1.0  48784.39304   48784.39304   
864         unknown            Алматы       5.0  48725.42295  243627.11480   
865         unknown            Алматы       2.0  31031.65229   62063.30457   
...             ...               ...       ...          ...           ...   
39961       unknown            Актобе       1.0  48756.92071   48756.92071   
39962       unknown            Актобе       1.0  48756.92071   48756.92071   
39963       unknown            Актобе       1.0  48756.92071   48756.92071   
39964       unknown            Актобе       1.0  48756.92071   48756.92071   
39965       unknown            Актобе       1.0  48756.92071   48756.92071   

                               oblact                  segment  \
861                          г.Алматы  Коммерческие автомобили   
862    Восточно-Казахстанская область  Коммерческие автомобили   
863               Акмолинская область  Коммерческие автомобили   
864                          г.Алматы  Коммерческие автомобили   
865                          г.Алматы  Коммерческие автомобили   
...                               ...                      ...   
39961             Актюбинская область  Коммерческие автомобили   
39962             Актюбинская область  Коммерческие автомобили   
39963             Актюбинская область  Коммерческие автомобили   
39964             Актюбинская область  Коммерческие автомобили   
39965             Актюбинская область  Коммерческие автомобили   

                           class  
861    Крупнотоннажные грузовики  
862    Крупнотоннажные грузовики  
863    Крупнотоннажные грузовики  
864    Крупнотоннажные грузовики  
865    Крупнотоннажные грузовики  
...                          ...  
39961  Крупнотоннажные грузовики  
39962  Крупнотоннажные грузовики  
39963  Крупнотоннажные грузовики  
39964  Крупнотоннажные грузовики  
39965  Крупнотоннажные грузовики  

[3980 rows x 18 columns]

In [ ]:
Астана
# 4. Группируем редкие страны (<1%) в 'другие'
country_freq = df['country'].value_counts(normalize=True)
rare_countries = country_freq[country_freq < 0.01].index


In [ ]:
df['country'] = df['country'].replace(rare_countries, 'другие')


In [ ]:
# 5. Красивый формат
df['country'] = df['country'].str.title()

In [ ]:
comment_stats = df['country'].value_counts(dropna=False).to_frame(name='Количество')

In [ ]:
comment_stats['Процент (%)'] = (comment_stats['Количество'] / len(df) * 100).round(2)

In [ ]:
print("\nСтатистика по 'country':")
print(comment_stats.head(50))